src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,458 @@
+   * @param data - The data to be put in the file.
+   * @param allowRecursiveCreation - Whether necessary but not existing intermediate containers may be created.
+   * @param metadata - Optional metadata that will be stored at `path/resourceName.metadata` if set.
+   *
+   * @returns Promise of the identifier of the newly created resource.
+   */
+  private async createFile(path: string, resourceName: string, data: Readable,
+    allowRecursiveCreation: boolean, metadata?: Readable): Promise<ResourceIdentifier> {
+    if (allowRecursiveCreation) {
+      await this.createContainer(path, '', true);
+    }
+    return new Promise(async(resolve, reject): Promise<any> => {
+      try {
+        const stats = await fsPromises.lstat(path);
+        if (!stats.isDirectory()) {
+          reject(new MethodNotAllowedHttpError('The given path is not a valid container.'));
+        } else {
+          if (metadata) {
+            this.createMetadata(`${path}${resourceName}.metadata`, metadata, reject);
+          }
+          const writeStream = createWriteStream(path + resourceName);
+
+          data.on('data', (chunk): any => writeStream.write(chunk));
+          data.on('error', reject);
+          data.on('end', (): any => writeStream.end());
+
+          writeStream.on('error', reject);
+          writeStream.on('finish', (): any => resolve({ path: this.mapFilepathToUrl(path + resourceName) }));
+        }
+      } catch (error) {
+        reject(new MethodNotAllowedHttpError());
+      }
+    });
+  }
+
+  /**
+   * Create a directory to represent a container.
+   * @param path - The path to the parent directory in which the new directory should be created.
+   * @param _containerName - The name of the directory to be created.
+   * @param allowRecursiveCreation - Whether necessary but not existing intermediate containers may be created.
+   * @param metadata - Optional metadata that will be stored at `path/_containerName/.metadata` if set.
+   *
+   * @returns Promise of the identifier of the newly created container.
+   */
+  private async createContainer(path: string, _containerName: string,
+    allowRecursiveCreation: boolean, metadata?: Readable): Promise<ResourceIdentifier> {
+    const containerName = ensureTrailingSlash(_containerName);
+    return new Promise(async(resolve, reject): Promise<any> => {
+      try {
+        if (allowRecursiveCreation) {
+          await fsPromises.mkdir(path + containerName, { recursive: true });
+          if (metadata) {
+            this.createMetadata(`${path}${containerName}.metadata`, metadata, reject);
+          }
+          resolve({ path: this.mapFilepathToUrl(path + containerName) });
+        } else {
+          const stats = await fsPromises.lstat(path);
+          if (!stats.isDirectory()) {
+            reject(new MethodNotAllowedHttpError('The given path is not a valid container.'));
+          } else {
+            await fsPromises.mkdir(path + containerName, { recursive: false });
```

Comment:
There is much overlap between this block and the one above (even more if you do `recursive: allowRecursiveCreation`) so this can probably be simplified.